PyInstaller/hooks/pre_find_module_path/hook-pyi_splash.py

```diff
@@ -0,0 +1,28 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2005-2019, PyInstaller Development Team.
+#
+# Distributed under the terms of the GNU General Public License with exception
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# -----------------------------------------------------------------------------
+
+"""
+This hook does not move a module that can be installed by a package manager,
+but points to a PyInstaller internal module that can be imported into
+the users python instance.
+
+The module is implemented in 'PyInstaller/fake-modules/pyi_splash.py'.
+"""
+
+import os
+
+from PyInstaller import PACKAGEPATH
+from PyInstaller.utils.hooks import logger
+
+
+def pre_find_module_path(api):
+    module_dir = os.path.join(PACKAGEPATH, 'fake-modules')
+
+    api.search_dirs = [module_dir]
+    logger.info('Adding pyi_splash module to application dependencies')
```

Comment:
Maybe before the module is redirected, check if a module with the same name is installed and if so, raise a SystemExit (or more descriptive Exception) ? Or rename the module to _pyi_splash? After all, it should not happen that the module is accidentally included although another one was meant.

Reply:
I had initially implemented the module as a runtime hook, but then the following problems occurred:

- If it should be implemented as a runtime hook (i.e. in `loader/rthooks`), then modulegraph must find a module named `pyi_splash` during analysis, otherwise the runtime hook will not be included. So a fake module must be imported. During the boot process the runtime hook is then executed, which creates a module and injects it into `sys.modules` so that it can be imported. The runtime hook would then look somewhat like this:
  ```python3
  import sys

  is_win = sys.platform.startswith("win")

  MSG_FINISH_TYPE = b'c'
  MSG_UPDATE_TYPE = b'u'

  # A class of type module? Weird
  # type(sys) == type 'module'
  class _PYISplashModule(type(sys)):

      def __init__(self):
          import os  # Modules need to be imported where
                     # they are used, not globally
          # Only super() syntax works
          super(PYISplashModule, self).__init__('pyi_splash')

          self._initialized = False
          self._pipe_write_file = None

          ...

      # these methods cannot be found by e.g. pydoc
      def close(self):
          ...

      def update_text(self, msg):
          ...
  
  
  sys.modules['pyi_splash'] = _PYISplashModule()
  ```
  I just think that this code is more of a workaround than a reasonable solution and may have dangers that are not obvious. Also, a runtime hook will overwrite a module that is actually already implemented with the fake-module, so one can simplify it to the current solution. Unfortunately, one cannot pass the runtime hook via the target splash, because then there is the danger that the runtime hook is behind the user's script and therefore never executed.

- If it is to be implemented as a loader module (under `loader/`), it will always be executed, regardless of its position in the CArchive before the user's script (since modules are executed before scripts), but there are no non-builtin modules available at that time that pyi_splash needs (such as os, atexit). Again, a workaround can be used there, but it makes the code ugly and not maintainable.

So I chose this method.
The hook under `hooks/hook-pyi_splash.py` can be removed if Python 2 support should be completely removed, because it was only necessary for the dependencies under Python 2.